File: solutions/java/src/main/java/com/epi/SymmetricBinaryTree.java
package com.epi;

import com.epi.BinaryTreePrototypeTemplate.BinaryTreeNode;

public class SymmetricBinaryTree {
  // @include
  public static boolean isSymmetric(BinaryTreeNode<Integer> tree) {
    return tree == null || checkSymmetric(tree.left, tree.right);
  }

  private static boolean checkSymmetric(BinaryTreeNode<Integer> subtree0,
                                        BinaryTreeNode<Integer> subtree1) {
    if (subtree0 == null && subtree1 == null) {
      return true;
    } else if (subtree0 != null && subtree1 != null) {
      return subtree0.data == subtree1.data
          && checkSymmetric(subtree0.left, subtree1.right)
          && checkSymmetric(subtree0.right, subtree1.left);
    }
    // One subtree is empty, and the other is not.
    return false;
  }
  // @exclude

  private static void simpleTest() {
    BinaryTreeNode<Integer> symmTree = new BinaryTreeNode<>();
    assert(isSymmetric(symmTree));
    symmTree.left = new BinaryTreeNode<Integer>();
    assert(!isSymmetric(symmTree));
    symmTree.right = new BinaryTreeNode<Integer>();
    assert(isSymmetric(symmTree));
    symmTree.right.right = new BinaryTreeNode<Integer>();
    assert(!isSymmetric(symmTree));
  }

  public static void main(String[] args) {
    simpleTest();
    // Non symmetric tree test
    //    x
    //  x   x
    // x   x x
    BinaryTreeNode<Integer> nonSymmTree = new BinaryTreeNode<>();
    nonSymmTree.left = new BinaryTreeNode<Integer>();
    nonSymmTree.left.left = new BinaryTreeNode<Integer>();
    nonSymmTree.right = new BinaryTreeNode<Integer>();
    nonSymmTree.right.left = new BinaryTreeNode<Integer>();
    nonSymmTree.right.right = new BinaryTreeNode<Integer>();
    assert(!isSymmetric(nonSymmTree));
    System.out.println(isSymmetric(nonSymmTree));
    // Symmetric tree test
    BinaryTreeNode<Integer> symmTree = new BinaryTreeNode<>();
    symmTree.left = new BinaryTreeNode<Integer>();
    symmTree.right = new BinaryTreeNode<Integer>();
    assert(isSymmetric(symmTree));
    System.out.println(isSymmetric(symmTree));
  }
}

File: solutions/cpp/Symmetric_binary_tree.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>

#include "./Binary_tree_prototype.h"

using std::boolalpha;
using std::cout;
using std::endl;
using std::unique_ptr;

bool CheckSymmetric(const unique_ptr<BinaryTreeNode<int>>&,
                       const unique_ptr<BinaryTreeNode<int>>&);

// @include
bool IsSymmetric(const unique_ptr<BinaryTreeNode<int>>& tree) {
  return tree == nullptr || CheckSymmetric(tree->left, tree->right);
}

bool CheckSymmetric(const unique_ptr<BinaryTreeNode<int>>& subtree_0,
                       const unique_ptr<BinaryTreeNode<int>>& subtree_1) {
  if (subtree_0 == nullptr && subtree_1 == nullptr) {
    return true;
  } else if (subtree_0 != nullptr && subtree_1 != nullptr) {
    return subtree_0->data == subtree_1->data &&
           CheckSymmetric(subtree_0->left, subtree_1->right) &&
           CheckSymmetric(subtree_0->right, subtree_1->left);
  }
  // One subtree is empty, and the other is not.
  return false;
}
// @exclude

int main(int argc, char* argv[]) {
  // Non symmetric tree test.
  //      3
  //    2   5
  //  1    4 6
  auto non_symm_root =
      unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  non_symm_root->left =
      unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  non_symm_root->left->left =
      unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  non_symm_root->right =
      unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  non_symm_root->right->left =
      unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  non_symm_root->right->right =
      unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  assert(!IsSymmetric(non_symm_root));
  cout << boolalpha << IsSymmetric(non_symm_root) << endl;
  // Symmetric tree test.
  unique_ptr<BinaryTreeNode<int>> symm_root =
      unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  symm_root->left = unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  symm_root->right = unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  assert(IsSymmetric(symm_root) == true);
  cout << boolalpha << IsSymmetric(symm_root) << endl;
  // Empty tree test.
  symm_root = nullptr;
  assert(IsSymmetric(symm_root) == true);
  cout << boolalpha << IsSymmetric(symm_root) << endl;
  return 0;
}
